server/mediaserver.go

```diff
@@ -412,8 +412,8 @@ func (s *LivepeerServer) registerConnection(rtmpStrm stream.RTMPVideoStream) (*r
 	hlsStrmID := core.MakeStreamID(mid, &vProfile)
 	s.connectionLock.Lock()
 	_, exists := s.rtmpConnections[mid]
-	s.connectionLock.Unlock()
```

Comment:
The following few lines of code can be very expensive, especially in other function call-outs where we don't necessarily always know how long they will take (eg, `NewBasicPlaylistManager` and `NewSessionManager`).

In fact, we all spent a few days debugging a deadlock in this exact same spot a few months ago because we were holding the lock for too long during `NewSessionManager`. We have a few options here:

* Make any external calls outside the lock. This is what the old code does.

* Check again that the session exists before assigning within the lock. We could add this.

* Accept that we might get multiple sessions or requests for a given manifest ID. This usually means the client is behaving inappropriately and shouldn't happen under normal conditions. In our case, the worst that happens is that "old" sessions get overwritten. That seems OK here. What we should ensure is that crash-prone bits are made more robust. If understand the root of the crash in https://github.com/livepeer/go-livepeer/pull/1284 correctly, it was something like this:

1. Multiple pushes come in very quickly such that the several sessions are created for the same manifest ID.
2. Multiple cleanup tickers are started
3. Cleanup ticker didn't check for the existence of the connection
4. Hence, if one connection was cleaned up, the second would lead to a crash

The actual solution for here seems to be to fix the cleanup ticker to check for existence of the connection, rather than making changes around locking here.

Reply:
> Check again that the session exists before assigning within the lock. We could add this.
> The actual solution for here seems to be to fix the cleanup ticker to check for existence of the connection

Made both 26a4773e393a7e8159e84dac30145ffe5482ffe7